base-trezor-keyring.js

```diff
@@ -0,0 +1,562 @@
+
+const hdPathString = `m/44'/60'/0'/0`;
+const SLIP0044TestnetPath = `m/44'/1'/0'/0`;
+
+const ALLOWED_HD_PATHS = {
+  [hdPathString]: true,
+  [SLIP0044TestnetPath]: true,
+};
+
+const keyringType = 'Trezor Hardware';
+const pathBase = 'm';
+const MAX_INDEX = 1000;
+const DELAY_BETWEEN_POPUPS = 1000;
+const TREZOR_CONNECT_MANIFEST = {
+  email: 'support@metamask.io',
+  appUrl: 'https://metamask.io',
+};
+
+function wait(ms) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+/**
+ * @typedef {import('@ethereumjs/tx').TypedTransaction} TypedTransaction
+ * @typedef {InstanceType<import("ethereumjs-tx")>} OldEthJsTransaction
+ */
+
+/**
+ * Check if the given transaction is made with ethereumjs-tx or @ethereumjs/tx
+ *
+ * Transactions built with older versions of ethereumjs-tx have a
+ * getChainId method that newer versions do not.
+ * Older versions are mutable
+ * while newer versions default to being immutable.
+ * Expected shape and type
+ * of data for v, r and s differ (Buffer (old) vs BN (new)).
+ *
+ * @param {TypedTransaction | OldEthJsTransaction} tx
+ * @returns {tx is OldEthJsTransaction} Returns `true` if tx is an old-style ethereumjs-tx transaction.
+ */
+function isOldStyleEthereumjsTx(tx) {
+  return typeof tx.getChainId === 'function';
+}
+
+class BaseTrezorKeyring extends EventEmitter {
+  constructor(opts = {}) {
+    super();
+    this.type = keyringType;
+    this.accounts = [];
+    this.hdk = new HDKey();
+    this.page = 0;
+    this.perPage = 5;
+    this.unlockedAccount = 0;
+    this.paths = {};
+    this.deserialize(opts);
+  }
+
+  /**
+   * Performs any asynchronous initialisation
+   *
+   * @returns {Promise<void>}
+   */
+  init() {
+    throw new Error('Method init not implemented');
+  }
+
+  /**
+   * Gets the model, if known.
+   * This may be `undefined` if the model hasn't been loaded yet.
+   *
+   * @returns {"T" | "1" | undefined}
+   */
+  getModel() {
+    return this.model;
+  }
+
+  /**
+   * Clears the iframe created by Trezor
+   *
+   * @returns {Promise<void>}
+   */
+  dispose() {
+    throw new Error('Method dispose not implemented');
+  }
+
+  /**
+   * Gets public key from device
+   *
+   * @returns {Promise}
+   */
+  _getPublicKey(_payload) {
+    throw new Error('Method _getPublicKey not implemented');
+  }
+
+  /**
+   * Signs transaction using the device
+   *
+   * @returns {Promise}
+   */
+  _ethereumSignTransaction(_payload) {
+    throw new Error('Method _ethereumSignTransaction not implemented');
+  }
+
+  /**
+   * Signs message using the device
+   *
+   * @returns {Promise}
+   */
+  _ethereumSignMessage(_payload) {
+    throw new Error('Method _ethereumSignMessage not implemented');
+  }
+
+  /**
+   * Signs typed data using the device
+   *
+   * @returns {Promise}
+   */
+  _ethereumSignTypedData(_payload) {
+    throw new Error('Method _ethereumSignTypedData not implemented');
+  }
+
+  serialize() {
```

Comment:
Nit: There is no serialization/deserialization happening so could be confusing. Maybe something like `getOptions()`/`loadOptions(opts)`, or even a unified `options(opts)`?

Reply:
I don't think we can't change the names for `serialize` and `deserialize`, they are common to all the keyrings and they are being called by `KeyringController` and the extension itself in multiple places.

Reply:
ACK!